packages/core/src/modules/discover-features/DiscoverFeaturesApi.ts

```diff
@@ -1,101 +1,144 @@
+    this.serviceMap = [v1Service, v2Service].reduce(
+      (serviceMap, service) => ({
+        ...serviceMap,
+        [service.version]: service,
+      }),
+      {}
+    ) as ServiceMap<DFSs>
   }
 
-  public async isProtocolSupported(connectionId: string, message: { type: ParsedMessageType }) {
-    const { protocolUri } = message.type
-
-    // Listen for response to our feature query
-    const replaySubject = new ReplaySubject(1)
-    this.eventEmitter
-      .observable<AgentMessageProcessedEvent>(AgentEventTypes.AgentMessageProcessed)
-      .pipe(
-        // Stop when the agent shuts down
-        takeUntil(this.stop$),
-        filterContextCorrelationId(this.agentContext.contextCorrelationId),
-        // filter by connection id and query disclose message type
-        filter(
-          (e) =>
-            e.payload.connection?.id === connectionId &&
-            canHandleMessageType(DiscloseMessage, parseMessageType(e.payload.message.type))
-        ),
-        // Return whether the protocol is supported
-        map((e) => {
-          const message = e.payload.message as DiscloseMessage
-          return message.protocols.map((p) => p.protocolId).includes(protocolUri)
-        }),
-        // TODO: make configurable
-        // If we don't have an answer in 7 seconds (no response, not supported, etc...) error
-        timeout(7000),
-        // We want to return false if an error occurred
-        catchError(() => of(false))
-      )
-      .subscribe(replaySubject)
-
-    await this.queryFeatures(connectionId, {
-      query: protocolUri,
-      comment: 'Detect if protocol is supported',
-    })
+  public getService<PVT extends DiscoverFeaturesService['version']>(protocolVersion: PVT): DiscoverFeaturesService {
+    if (!this.serviceMap[protocolVersion]) {
+      throw new AriesFrameworkError(`No discover features service registered for protocol version ${protocolVersion}`)
+    }
 
-    const isProtocolSupported = await firstValueFrom(replaySubject)
-    return isProtocolSupported
+    return this.serviceMap[protocolVersion]
   }
 
-  public async queryFeatures(connectionId: string, options: { query: string; comment?: string }) {
-    const connection = await this.connectionService.getById(this.agentContext, connectionId)
+  /**
+   * Send a query to an existing connection for discovering supported features of any kind. If desired, do the query synchronously,
+   * meaning that it will await the response (or timeout) before resolving. Otherwise, response can be hooked by subscribing to
+   * {DiscoverFeaturesDisclosureReceivedEvent}.
+   *
+   * Note: V1 protocol only supports a single query and is limited to protocols
+   *
+   * @param options feature queries to perform, protocol version and optional comment string (only used
+   * in V1 protocol). If awaitDisclosures is set, perform the query synchronously with awaitDisclosuresTimeoutMs timeout.
+   */
+  public async queryFeatures(options: QueryFeaturesOptions<DFSs>) {
```

Comment:
Maybe too much overhead/complex but do you think we can add return type definition that takes into account the answer to awaitDosclosures?

That way you don't have an optional return type, which can improve DX